File: waitwise/utils/calculations.py
import pandas as pd

from utils.scenarios import COSTING_METHOD_MAP, SCENARIO_MAP, TARGETING_MODE_MAP


def safe_divide(numerator: float, denominator: float) -> float:
    if denominator == 0:
        return 0.0
    return numerator / denominator


def clamp_rate(value: float) -> float:
    return max(0.0, min(1.0, value))


def get_discount_factor(year: int, discount_rate: float) -> float:
    return 1 / ((1 + discount_rate) ** (year - 1))


def get_targeting_adjustments(inputs: dict) -> dict:
    targeting = TARGETING_MODE_MAP[inputs["targeting_mode"]]
    adjusted_waiting_list = inputs["starting_waiting_list_size"] * targeting["population_multiplier"]
    adjusted_reach_rate = clamp_rate(inputs["intervention_reach_rate"] * targeting["reach_multiplier"])
    adjusted_escalation_rate = clamp_rate(inputs["monthly_escalation_rate"] * targeting["risk_multiplier"])

    return {
        "adjusted_waiting_list": adjusted_waiting_list,
        "adjusted_reach_rate": adjusted_reach_rate,
        "adjusted_escalation_rate": adjusted_escalation_rate,
    }


def calculate_gross_savings(
    escalations_avoided: float,
    admissions_avoided: float,
    bed_days_avoided: float,
    inputs: dict,
) -> float:
    costing = COSTING_METHOD_MAP[inputs["costing_method"]]
    escalation_and_admission_savings = (
        escalations_avoided * inputs["cost_per_escalation"]
        + admissions_avoided * inputs["cost_per_admission"]
    )
    bed_day_value = bed_days_avoided * inputs["cost_per_bed_day"]

    if costing["mode"] == "acute":
        return escalation_and_admission_savings
    if costing["mode"] == "bed_day":
        return bed_day_value
    return escalation_and_admission_savings + bed_day_value


def _run_model_core(inputs: dict) -> dict:
    targeting = get_targeting_adjustments(inputs)

    starting_waiting_list = targeting["adjusted_waiting_list"]
    monthly_inflow = inputs["monthly_inflow"]
    monthly_throughput = inputs["baseline_monthly_throughput"]

    yearly_rows = []
    cumulative_programme_cost = 0.0
    cumulative_gross_savings = 0.0
    cumulative_net_cost = 0.0
    waiting_list_current = starting_waiting_list

    for year in range(1, inputs["time_horizon_years"] + 1):
        effect_multiplier = (1 - inputs["effect_decay_rate"]) ** (year - 1)
        participation_multiplier = (1 - inputs["participation_dropoff_rate"]) ** (year - 1)

        effective_reach = targeting["adjusted_reach_rate"] * participation_multiplier
        demand_reduction = inputs["demand_reduction_effect"] * effect_multiplier
        throughput_increase = inputs["throughput_increase_effect"] * effect_multiplier
        escalation_reduction = inputs["escalation_reduction_effect"] * effect_multiplier

        annual_inflow = monthly_inflow * 12
        annual_baseline_throughput = monthly_throughput * 12

        reduced_inflow = annual_inflow * (1 - (demand_reduction * effective_reach))
        improved_throughput = annual_baseline_throughput * (1 + (throughput_increase * effective_reach))

        waiting_list_next = max(waiting_list_current + reduced_inflow - improved_throughput, 0)
        baseline_waiting_list_next = max(waiting_list_current + annual_inflow - annual_baseline_throughput, 0)

        waiting_list_reduction = max(baseline_waiting_list_next - waiting_list_next, 0)

        annual_escalations_baseline = waiting_list_current * targeting["adjusted_escalation_rate"] * 12
        escalations_avoided = annual_escalations_baseline * escalation_reduction * effective_reach
        admissions_avoided = escalations_avoided * inputs["admission_rate_after_escalation"]
        bed_days_avoided = admissions_avoided * inputs["average_length_of_stay"]

        patients_reached = waiting_list_current * effective_reach
        programme_cost = patients_reached * inputs["intervention_cost_per_patient_reached"]
        gross_savings = calculate_gross_savings(
            escalations_avoided,
            admissions_avoided,
            bed_days_avoided,
            inputs,
        )
        net_cost = programme_cost - gross_savings
        qalys_gained = escalations_avoided * inputs["qaly_gain_per_escalation_avoided"]

        discount_factor = get_discount_factor(year, inputs["discount_rate"])
        discounted_programme_cost = programme_cost * discount_factor
        discounted_gross_savings = gross_savings * discount_factor
        discounted_net_cost = net_cost * discount_factor
        discounted_qalys = qalys_gained * discount_factor

        cumulative_programme_cost += programme_cost
        cumulative_gross_savings += gross_savings
        cumulative_net_cost += net_cost

        yearly_rows.append(
            {
                "year": year,
                "waiting_list_start": waiting_list_current,
                "waiting_list_end": waiting_list_next,
                "waiting_list_reduction": waiting_list_reduction,
                "escalations_avoided": escalations_avoided,
                "admissions_avoided": admissions_avoided,
                "bed_days_avoided": bed_days_avoided,
                "patients_reached": patients_reached,
                "programme_cost": programme_cost,
                "gross_savings": gross_savings,
                "net_cost": net_cost,
                "qalys_gained": qalys_gained,
                "discount_factor": discount_factor,
                "discounted_programme_cost": discounted_programme_cost,
                "discounted_gross_savings": discounted_gross_savings,
                "discounted_net_cost": discounted_net_cost,
                "discounted_qalys": discounted_qalys,
                "cumulative_programme_cost": cumulative_programme_cost,
                "cumulative_gross_savings": cumulative_gross_savings,
                "cumulative_net_cost": cumulative_net_cost,
            }
        )

        waiting_list_current = waiting_list_next

    yearly_results = pd.DataFrame(yearly_rows)

    waiting_list_reduction_total = yearly_results["waiting_list_reduction"].sum()
    escalations_avoided_total = yearly_results["escalations_avoided"].sum()
    admissions_avoided_total = yearly_results["admissions_avoided"].sum()
    bed_days_avoided_total = yearly_results["bed_days_avoided"].sum()
    programme_cost_total = yearly_results["programme_cost"].sum()
    gross_savings_total = yearly_results["gross_savings"].sum()
    discounted_programme_cost_total = yearly_results["discounted_programme_cost"].sum()
    discounted_gross_savings_total = yearly_results["discounted_gross_savings"].sum()
    discounted_net_cost_total = yearly_results["discounted_net_cost"].sum()
    discounted_qalys_total = yearly_results["discounted_qalys"].sum()

    discounted_cost_per_qaly = safe_divide(discounted_net_cost_total, discounted_qalys_total)
    roi = safe_divide(gross_savings_total, programme_cost_total)

    return {
        "waiting_list_start_year_1": yearly_results.iloc[0]["waiting_list_start"],
        "waiting_list_end_final": yearly_results.iloc[-1]["waiting_list_end"],
        "waiting_list_reduction_total": waiting_list_reduction_total,
        "escalations_avoided_total": escalations_avoided_total,
        "admissions_avoided_total": admissions_avoided_total,
        "bed_days_avoided_total": bed_days_avoided_total,
        "programme_cost_total": programme_cost_total,
        "gross_savings_total": gross_savings_total,
        "discounted_programme_cost_total": discounted_programme_cost_total,
        "discounted_gross_savings_total": discounted_gross_savings_total,
        "discounted_net_cost_total": discounted_net_cost_total,
        "discounted_qalys_total": discounted_qalys_total,
        "discounted_cost_per_qaly": discounted_cost_per_qaly,
        "roi": roi,
        "yearly_results": yearly_results,
    }


def run_model(inputs: dict) -> dict:
    core = _run_model_core(inputs)
    core.update(
        {
            "break_even_effect_required": clamp_rate(calculate_break_even_effect(inputs)),
            "break_even_cost_per_patient": calculate_break_even_cost_per_patient(inputs),
            "break_even_horizon": calculate_break_even_horizon(inputs, max_years=10),
        }
    )
    return core


def calculate_break_even_effect(inputs: dict) -> float:
    targeting = get_targeting_adjustments(inputs)
    waiting_list_base = targeting["adjusted_waiting_list"]

    numerator = 0.0
    denominator = 0.0

    for year in range(1, inputs["time_horizon_years"] + 1):
        participation_multiplier = (1 - inputs["participation_dropoff_rate"]) ** (year - 1)
        effect_multiplier = (1 - inputs["effect_decay_rate"]) ** (year - 1)
        effective_reach = targeting["adjusted_reach_rate"] * participation_multiplier

        patients_reached = waiting_list_base * effective_reach
        escalations_avoided_per_unit = (
            waiting_list_base
            * targeting["adjusted_escalation_rate"]
            * 12
            * effect_multiplier
            * effective_reach
        )
        admissions_avoided_per_unit = escalations_avoided_per_unit * inputs["admission_rate_after_escalation"]
        bed_days_avoided_per_unit = admissions_avoided_per_unit * inputs["average_length_of_stay"]

        gross_savings_per_unit = calculate_gross_savings(
            escalations_avoided_per_unit,
            admissions_avoided_per_unit,
            bed_days_avoided_per_unit,
            inputs,
        )
        qaly_value_per_unit = (
            escalations_avoided_per_unit
            * inputs["qaly_gain_per_escalation_avoided"]
            * inputs["cost_effectiveness_threshold"]
        )

        discount_factor = get_discount_factor(year, inputs["discount_rate"])
        numerator += patients_reached * inputs["intervention_cost_per_patient_reached"] * discount_factor
        denominator += (gross_savings_per_unit + qaly_value_per_unit) * discount_factor

    return safe_divide(numerator, denominator)


def calculate_break_even_cost_per_patient(inputs: dict) -> float:
    targeting = get_targeting_adjustments(inputs)
    waiting_list_base = targeting["adjusted_waiting_list"]

    total_discounted_patients_reached = 0.0
    total_discounted_value = 0.0

    for year in range(1, inputs["time_horizon_years"] + 1):
        participation_multiplier = (1 - inputs["participation_dropoff_rate"]) ** (year - 1)
        effect_multiplier = (1 - inputs["effect_decay_rate"]) ** (year - 1)
        effective_reach = targeting["adjusted_reach_rate"] * participation_multiplier

        patients_reached = waiting_list_base * effective_reach
        blended_effect = (
            inputs["demand_reduction_effect"]
            + inputs["throughput_increase_effect"]
            + inputs["escalation_reduction_effect"]
        ) / 3

        escalations_avoided = (
            waiting_list_base
            * targeting["adjusted_escalation_rate"]
            * 12
            * blended_effect
            * effect_multiplier
            * effective_reach
        )
        admissions_avoided = escalations_avoided * inputs["admission_rate_after_escalation"]
        bed_days_avoided = admissions_avoided * inputs["average_length_of_stay"]

        gross_savings = calculate_gross_savings(
            escalations_avoided,
            admissions_avoided,
            bed_days_avoided,
            inputs,
        )
        qaly_value = (
            escalations_avoided
            * inputs["qaly_gain_per_escalation_avoided"]
            * inputs["cost_effectiveness_threshold"]
        )

        discount_factor = get_discount_factor(year, inputs["discount_rate"])
        total_discounted_patients_reached += patients_reached * discount_factor
        total_discounted_value += (gross_savings + qaly_value) * discount_factor

    return safe_divide(total_discounted_value, total_discounted_patients_reached)


def calculate_break_even_horizon(inputs: dict, max_years: int = 10) -> str:
    for horizon in range(1, max_years + 1):
        test_inputs = inputs.copy()
        test_inputs["time_horizon_years"] = horizon
        result = _run_model_core(test_inputs)
        if (
            result["discounted_cost_per_qaly"] > 0
            and result["discounted_cost_per_qaly"] <= inputs["cost_effectiveness_threshold"]
        ):
            return f"{horizon} year{'s' if horizon != 1 else ''}"
        if result["discounted_net_cost_total"] < 0:
            return f"{horizon} year{'s' if horizon != 1 else ''}"

    return f">{max_years} years"


def run_bounded_uncertainty(inputs: dict) -> pd.DataFrame:
    cases = {
        "Low": {
            "demand_reduction_effect": clamp_rate(inputs["demand_reduction_effect"] * 0.8),
            "throughput_increase_effect": clamp_rate(inputs["throughput_increase_effect"] * 0.8),
            "escalation_reduction_effect": clamp_rate(inputs["escalation_reduction_effect"] * 0.8),
            "intervention_cost_per_patient_reached": inputs["intervention_cost_per_patient_reached"] * 1.2,
            "qaly_gain_per_escalation_avoided": inputs["qaly_gain_per_escalation_avoided"] * 0.8,
            "participation_dropoff_rate": clamp_rate(inputs["participation_dropoff_rate"] * 1.2),
            "dominant_domain": "Delivery assumptions",
        },
        "Base": {
            "dominant_domain": "Base case",
        },
        "High": {
            "demand_reduction_effect": clamp_rate(inputs["demand_reduction_effect"] * 1.2),
            "throughput_increase_effect": clamp_rate(inputs["throughput_increase_effect"] * 1.2),
            "escalation_reduction_effect": clamp_rate(inputs["escalation_reduction_effect"] * 1.2),
            "intervention_cost_per_patient_reached": inputs["intervention_cost_per_patient_reached"] * 0.8,
            "qaly_gain_per_escalation_avoided": inputs["qaly_gain_per_escalation_avoided"] * 1.2,
            "participation_dropoff_rate": clamp_rate(inputs["participation_dropoff_rate"] * 0.8),
            "dominant_domain": "Clinical and delivery assumptions",
        },
    }

    rows = []
    for case_name, overrides in cases.items():
        case_inputs = inputs.copy()
        dominant_domain = overrides.pop("dominant_domain")
        case_inputs.update(overrides)
        result = _run_model_core(case_inputs)

        if result["discounted_net_cost_total"] < 0:
            decision_status = "Appears cost-saving"
        elif 0 < result["discounted_cost_per_qaly"] <= inputs["cost_effectiveness_threshold"]:
            decision_status = "Appears cost-effective"
        else:
            decision_status = "Above current threshold"

        rows.append(
            {
                "case": case_name,
                "waiting_list_reduction_total": result["waiting_list_reduction_total"],
                "discounted_net_cost_total": result["discounted_net_cost_total"],
                "discounted_cost_per_qaly": result["discounted_cost_per_qaly"],
                "dominant_domain": dominant_domain,
                "decision_status": decision_status,
            }
        )

    return pd.DataFrame(rows)


def build_comparator_case(defaults: dict, base_inputs: dict, comparator_mode: str) -> dict:
    comparator_inputs = defaults.copy()

    if comparator_mode in SCENARIO_MAP:
        comparator_inputs.update(SCENARIO_MAP[comparator_mode](defaults))
    else:
        comparator_inputs.update(base_inputs)

    comparator_inputs["time_horizon_years"] = base_inputs["time_horizon_years"]
    comparator_inputs["discount_rate"] = base_inputs["discount_rate"]
    comparator_inputs["costing_method"] = base_inputs["costing_method"]
    comparator_inputs["cost_effectiveness_threshold"] = base_inputs["cost_effectiveness_threshold"]
    comparator_inputs["cost_per_escalation"] = base_inputs["cost_per_escalation"]
    comparator_inputs["cost_per_admission"] = base_inputs["cost_per_admission"]
    comparator_inputs["cost_per_bed_day"] = base_inputs["cost_per_bed_day"]
    comparator_inputs["qaly_gain_per_escalation_avoided"] = base_inputs["qaly_gain_per_escalation_avoided"]

    return comparator_inputs

File: site/src/lib/waitwise/calculations.ts
import { COSTING_METHOD_MAP, SCENARIO_MAP, TARGETING_MODE_MAP } from "@/lib/waitwise/scenarios";
import type {
  ComparatorOption,
  Inputs,
  ModelResults,
  UncertaintyRow,
  YearlyResultRow,
} from "@/lib/waitwise/types";

export function safeDivide(numerator: number, denominator: number): number {
  if (denominator === 0) {
    return 0;
  }
  return numerator / denominator;
}

export function clampRate(value: number): number {
  return Math.max(0, Math.min(1, value));
}

export function getDiscountFactor(year: number, discountRate: number): number {
  return 1 / (1 + discountRate) ** (year - 1);
}

export function getTargetingAdjustments(inputs: Inputs) {
  const targeting = TARGETING_MODE_MAP[inputs.targeting_mode];
  const adjustedWaitingList =
    inputs.starting_waiting_list_size * targeting.population_multiplier;
  const adjustedReachRate = clampRate(
    inputs.intervention_reach_rate * targeting.reach_multiplier,
  );
  const adjustedEscalationRate = clampRate(
    inputs.monthly_escalation_rate * targeting.risk_multiplier,
  );

  return {
    adjusted_waiting_list: adjustedWaitingList,
    adjusted_reach_rate: adjustedReachRate,
    adjusted_escalation_rate: adjustedEscalationRate,
  };
}

export function calculateGrossSavings(
  escalationsAvoided: number,
  admissionsAvoided: number,
  bedDaysAvoided: number,
  inputs: Inputs,
): number {
  const costing = COSTING_METHOD_MAP[inputs.costing_method];

  const escalationAndAdmissionSavings =
    escalationsAvoided * inputs.cost_per_escalation +
    admissionsAvoided * inputs.cost_per_admission;

  const bedDayValue = bedDaysAvoided * inputs.cost_per_bed_day;

  if (costing.mode === "acute") {
    return escalationAndAdmissionSavings;
  }

  if (costing.mode === "bed_day") {
    return bedDayValue;
  }

  return escalationAndAdmissionSavings + bedDayValue;
}

function runModelCore(inputs: Inputs): ModelResults {
  const targeting = getTargetingAdjustments(inputs);

  const startingWaitingList = targeting.adjusted_waiting_list;
  const monthlyInflow = inputs.monthly_inflow;
  const monthlyThroughput = inputs.baseline_monthly_throughput;

  const yearlyRows: YearlyResultRow[] = [];
  let cumulativeProgrammeCost = 0;
  let cumulativeGrossSavings = 0;
  let cumulativeNetCost = 0;
  let waitingListCurrent = startingWaitingList;

  for (let year = 1; year <= inputs.time_horizon_years; year += 1) {
    const effectMultiplier = (1 - inputs.effect_decay_rate) ** (year - 1);
    const participationMultiplier =
      (1 - inputs.participation_dropoff_rate) ** (year - 1);

    const effectiveReach =
      targeting.adjusted_reach_rate * participationMultiplier;
    const demandReduction =
      inputs.demand_reduction_effect * effectMultiplier;
    const throughputIncrease =
      inputs.throughput_increase_effect * effectMultiplier;
    const escalationReduction =
      inputs.escalation_reduction_effect * effectMultiplier;

    const annualInflow = monthlyInflow * 12;
    const annualBaselineThroughput = monthlyThroughput * 12;

    const reducedInflow =
      annualInflow * (1 - demandReduction * effectiveReach);
    const improvedThroughput =
      annualBaselineThroughput * (1 + throughputIncrease * effectiveReach);

    const waitingListNext = Math.max(
      waitingListCurrent + reducedInflow - improvedThroughput,
      0,
    );
    const baselineWaitingListNext = Math.max(
      waitingListCurrent + annualInflow - annualBaselineThroughput,
      0,
    );

    const waitingListReduction = Math.max(
      baselineWaitingListNext - waitingListNext,
      0,
    );

    const annualEscalationsBaseline =
      waitingListCurrent * targeting.adjusted_escalation_rate * 12;
    const escalationsAvoided =
      annualEscalationsBaseline * escalationReduction * effectiveReach;
    const admissionsAvoided =
      escalationsAvoided * inputs.admission_rate_after_escalation;
    const bedDaysAvoided =
      admissionsAvoided * inputs.average_length_of_stay;

    const patientsReached = waitingListCurrent * effectiveReach;
    const programmeCost =
      patientsReached * inputs.intervention_cost_per_patient_reached;

    const grossSavings = calculateGrossSavings(
      escalationsAvoided,
      admissionsAvoided,
      bedDaysAvoided,
      inputs,
    );

    const netCost = programmeCost - grossSavings;
    const qalysGained =
      escalationsAvoided * inputs.qaly_gain_per_escalation_avoided;

    const discountFactor = getDiscountFactor(year, inputs.discount_rate);
    const discountedProgrammeCost = programmeCost * discountFactor;
    const discountedGrossSavings = grossSavings * discountFactor;
    const discountedNetCost = netCost * discountFactor;
    const discountedQalys = qalysGained * discountFactor;

    cumulativeProgrammeCost += programmeCost;
    cumulativeGrossSavings += grossSavings;
    cumulativeNetCost += netCost;

    yearlyRows.push({
      year,
      waiting_list_start: waitingListCurrent,
      waiting_list_end: waitingListNext,
      waiting_list_reduction: waitingListReduction,
      escalations_avoided: escalationsAvoided,
      admissions_avoided: admissionsAvoided,
      bed_days_avoided: bedDaysAvoided,
      patients_reached: patientsReached,
      programme_cost: programmeCost,
      gross_savings: grossSavings,
      net_cost: netCost,
      qalys_gained: qalysGained,
      discount_factor: discountFactor,
      discounted_programme_cost: discountedProgrammeCost,
      discounted_gross_savings: discountedGrossSavings,
      discounted_net_cost: discountedNetCost,
      discounted_qalys: discountedQalys,
      cumulative_programme_cost: cumulativeProgrammeCost,
      cumulative_gross_savings: cumulativeGrossSavings,
      cumulative_net_cost: cumulativeNetCost,
    });

    waitingListCurrent = waitingListNext;
  }

  const waitingListReductionTotal = yearlyRows.reduce(
    (sum, row) => sum + row.waiting_list_reduction,
    0,
  );
  const escalationsAvoidedTotal = yearlyRows.reduce(
    (sum, row) => sum + row.escalations_avoided,
    0,
  );
  const admissionsAvoidedTotal = yearlyRows.reduce(
    (sum, row) => sum + row.admissions_avoided,
    0,
  );
  const bedDaysAvoidedTotal = yearlyRows.reduce(
    (sum, row) => sum + row.bed_days_avoided,
    0,
  );
  const programmeCostTotal = yearlyRows.reduce(
    (sum, row) => sum + row.programme_cost,
    0,
  );
  const grossSavingsTotal = yearlyRows.reduce(
    (sum, row) => sum + row.gross_savings,
    0,
  );
  const discountedProgrammeCostTotal = yearlyRows.reduce(
    (sum, row) => sum + row.discounted_programme_cost,
    0,
  );
  const discountedGrossSavingsTotal = yearlyRows.reduce(
    (sum, row) => sum + row.discounted_gross_savings,
    0,
  );
  const discountedNetCostTotal = yearlyRows.reduce(
    (sum, row) => sum + row.discounted_net_cost,
    0,
  );
  const discountedQalysTotal = yearlyRows.reduce(
    (sum, row) => sum + row.discounted_qalys,
    0,
  );

  const discountedCostPerQaly = safeDivide(
    discountedNetCostTotal,
    discountedQalysTotal,
  );
  const roi = safeDivide(grossSavingsTotal, programmeCostTotal);

  return {
    waiting_list_start_year_1: yearlyRows[0]?.waiting_list_start ?? 0,
    waiting_list_end_final: yearlyRows.at(-1)?.waiting_list_end ?? 0,
    waiting_list_reduction_total: waitingListReductionTotal,
    escalations_avoided_total: escalationsAvoidedTotal,
    admissions_avoided_total: admissionsAvoidedTotal,
    bed_days_avoided_total: bedDaysAvoidedTotal,
    programme_cost_total: programmeCostTotal,
    gross_savings_total: grossSavingsTotal,
    discounted_programme_cost_total: discountedProgrammeCostTotal,
    discounted_gross_savings_total: discountedGrossSavingsTotal,
    discounted_net_cost_total: discountedNetCostTotal,
    discounted_qalys_total: discountedQalysTotal,
    discounted_cost_per_qaly: discountedCostPerQaly,
    roi,
    yearly_results: yearlyRows,
    break_even_effect_required: 0,
    break_even_cost_per_patient: 0,
    break_even_horizon: "",
  };
}

export function calculateBreakEvenEffect(inputs: Inputs): number {
  const targeting = getTargetingAdjustments(inputs);
  const waitingListBase = targeting.adjusted_waiting_list;

  let numerator = 0;
  let denominator = 0;

  for (let year = 1; year <= inputs.time_horizon_years; year += 1) {
    const participationMultiplier =
      (1 - inputs.participation_dropoff_rate) ** (year - 1);
    const effectMultiplier =
      (1 - inputs.effect_decay_rate) ** (year - 1);
    const effectiveReach =
      targeting.adjusted_reach_rate * participationMultiplier;

    const patientsReached = waitingListBase * effectiveReach;
    const escalationsAvoidedPerUnit =
      waitingListBase *
      targeting.adjusted_escalation_rate *
      12 *
      effectMultiplier *
      effectiveReach;

    const admissionsAvoidedPerUnit =
      escalationsAvoidedPerUnit * inputs.admission_rate_after_escalation;
    const bedDaysAvoidedPerUnit =
      admissionsAvoidedPerUnit * inputs.average_length_of_stay;

    const grossSavingsPerUnit = calculateGrossSavings(
      escalationsAvoidedPerUnit,
      admissionsAvoidedPerUnit,
      bedDaysAvoidedPerUnit,
      inputs,
    );

    const qalyValuePerUnit =
      escalationsAvoidedPerUnit *
      inputs.qaly_gain_per_escalation_avoided *
      inputs.cost_effectiveness_threshold;

    const discountFactor = getDiscountFactor(year, inputs.discount_rate);
    numerator +=
      patientsReached *
      inputs.intervention_cost_per_patient_reached *
      discountFactor;
    denominator += (grossSavingsPerUnit + qalyValuePerUnit) * discountFactor;
  }

  return safeDivide(numerator, denominator);
}

export function calculateBreakEvenCostPerPatient(inputs: Inputs): number {
  const targeting = getTargetingAdjustments(inputs);
  const waitingListBase = targeting.adjusted_waiting_list;

  let totalDiscountedPatientsReached = 0;
  let totalDiscountedValue = 0;

  for (let year = 1; year <= inputs.time_horizon_years; year += 1) {
    const participationMultiplier =
      (1 - inputs.participation_dropoff_rate) ** (year - 1);
    const effectMultiplier =
      (1 - inputs.effect_decay_rate) ** (year - 1);
    const effectiveReach =
      targeting.adjusted_reach_rate * participationMultiplier;

    const patientsReached = waitingListBase * effectiveReach;

    const blendedEffect =
      (inputs.demand_reduction_effect +
        inputs.throughput_increase_effect +
        inputs.escalation_reduction_effect) /
      3;

    const escalationsAvoided =
      waitingListBase *
      targeting.adjusted_escalation_rate *
      12 *
      blendedEffect *
      effectMultiplier *
      effectiveReach;

    const admissionsAvoided =
      escalationsAvoided * inputs.admission_rate_after_escalation;
    const bedDaysAvoided =
      admissionsAvoided * inputs.average_length_of_stay;

    const grossSavings = calculateGrossSavings(
      escalationsAvoided,
      admissionsAvoided,
      bedDaysAvoided,
      inputs,
    );

    const qalyValue =
      escalationsAvoided *
      inputs.qaly_gain_per_escalation_avoided *
      inputs.cost_effectiveness_threshold;

    const discountFactor = getDiscountFactor(year, inputs.discount_rate);
    totalDiscountedPatientsReached += patientsReached * discountFactor;
    totalDiscountedValue += (grossSavings + qalyValue) * discountFactor;
  }

  return safeDivide(totalDiscountedValue, totalDiscountedPatientsReached);
}

export function calculateBreakEvenHorizon(
  inputs: Inputs,
  maxYears = 10,
): string {
  for (let horizon = 1; horizon <= maxYears; horizon += 1) {
    const testInputs: Inputs = {
      ...inputs,
      time_horizon_years: horizon as Inputs["time_horizon_years"],
    };

    const result = runModelCore(testInputs);

    if (
      result.discounted_cost_per_qaly > 0 &&
      result.discounted_cost_per_qaly <= inputs.cost_effectiveness_threshold
    ) {
      return `${horizon} year${horizon === 1 ? "" : "s"}`;
    }

    if (result.discounted_net_cost_total < 0) {
      return `${horizon} year${horizon === 1 ? "" : "s"}`;
    }
  }

  return `>${maxYears} years`;
}

export function runModel(inputs: Inputs): ModelResults {
  const core = runModelCore(inputs);

  return {
    ...core,
    break_even_effect_required: clampRate(calculateBreakEvenEffect(inputs)),
    break_even_cost_per_patient: calculateBreakEvenCostPerPatient(inputs),
    break_even_horizon: calculateBreakEvenHorizon(inputs, 10),
  };
}

export function getDecisionStatusForResult(
  discountedNetCostTotal: number,
  discountedCostPerQaly: number,
  threshold: number,
): string {
  if (discountedNetCostTotal < 0) {
    return "Appears cost-saving";
  }

  if (discountedCostPerQaly > 0 && discountedCostPerQaly <= threshold) {
    return "Appears cost-effective";
  }

  return "Above current threshold";
}

export function runBoundedUncertainty(inputs: Inputs): UncertaintyRow[] {
  const cases = [
    {
      case: "Low",
      dominantDomain: "Delivery assumptions",
      overrides: {
        demand_reduction_effect: clampRate(inputs.demand_reduction_effect * 0.8),
        throughput_increase_effect: clampRate(
          inputs.throughput_increase_effect * 0.8,
        ),
        escalation_reduction_effect: clampRate(
          inputs.escalation_reduction_effect * 0.8,
        ),
        intervention_cost_per_patient_reached:
          inputs.intervention_cost_per_patient_reached * 1.2,
        qaly_gain_per_escalation_avoided:
          inputs.qaly_gain_per_escalation_avoided * 0.8,
        participation_dropoff_rate: clampRate(
          inputs.participation_dropoff_rate * 1.2,
        ),
      },
    },
    {
      case: "Base",
      dominantDomain: "Base case",
      overrides: {},
    },
    {
      case: "High",
      dominantDomain: "Clinical and delivery assumptions",
      overrides: {
        demand_reduction_effect: clampRate(inputs.demand_reduction_effect * 1.2),
        throughput_increase_effect: clampRate(
          inputs.throughput_increase_effect * 1.2,
        ),
        escalation_reduction_effect: clampRate(
          inputs.escalation_reduction_effect * 1.2,
        ),
        intervention_cost_per_patient_reached:
          inputs.intervention_cost_per_patient_reached * 0.8,
        qaly_gain_per_escalation_avoided:
          inputs.qaly_gain_per_escalation_avoided * 1.2,
        participation_dropoff_rate: clampRate(
          inputs.participation_dropoff_rate * 0.8,
        ),
      },
    },
  ] as const;

  return cases.map((caseConfig) => {
    const caseInputs: Inputs = {
      ...inputs,
      ...caseConfig.overrides,
    };

    const result = runModelCore(caseInputs);
    const decisionStatus = getDecisionStatusForResult(
      result.discounted_net_cost_total,
      result.discounted_cost_per_qaly,
      inputs.cost_effectiveness_threshold,
    );

    return {
      case: caseConfig.case,
      waiting_list_reduction_total: result.waiting_list_reduction_total,
      discounted_net_cost_total: result.discounted_net_cost_total,
      discounted_cost_per_qaly: result.discounted_cost_per_qaly,
      dominant_domain: caseConfig.dominantDomain,
      decision_status: decisionStatus,
    };
  });
}

export function buildComparatorCase(
  defaults: Inputs,
  baseInputs: Inputs,
  comparatorMode: ComparatorOption,
): Inputs {
  const comparatorInputs: Inputs = { ...defaults };

  const scenarioBuilder = SCENARIO_MAP[comparatorMode];
  if (scenarioBuilder) {
    Object.assign(comparatorInputs, scenarioBuilder(defaults));
  } else {
    Object.assign(comparatorInputs, baseInputs);
  }

  comparatorInputs.time_horizon_years = baseInputs.time_horizon_years;
  comparatorInputs.discount_rate = baseInputs.discount_rate;
  comparatorInputs.costing_method = baseInputs.costing_method;
  comparatorInputs.cost_effectiveness_threshold =
    baseInputs.cost_effectiveness_threshold;
  comparatorInputs.cost_per_escalation = baseInputs.cost_per_escalation;
  comparatorInputs.cost_per_admission = baseInputs.cost_per_admission;
  comparatorInputs.cost_per_bed_day = baseInputs.cost_per_bed_day;
  comparatorInputs.qaly_gain_per_escalation_avoided =
    baseInputs.qaly_gain_per_escalation_avoided;

  return comparatorInputs;
}